fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanArrearsAgingServiceImpl.java

```diff
@@ -423,18 +423,14 @@ public OriginalScheduleExtractor(final String loanIdsAsString) {
 
             while (rs.next()) {
                 Long loanId = rs.getLong("loanId");
-                List<LoanSchedulePeriodData> periodDatas = new ArrayList<>();
-                LoanSchedulePeriodData loanSchedulePeriodData = fetchLoanSchedulePeriodData(rs);
-                periodDatas.add(loanSchedulePeriodData);
-                while (rs.next()) {
-                    Long tempLoanId = rs.getLong("loanId");
-                    if (loanId.equals(tempLoanId)) {
```

Comment:
I'm trying to understand this to the best I can but I just have the feeling that this piece of logic is missing. Can you maybe make it a little clearer for me?
Thanks.

Reply:
As far as I can see, the logic is trying to build a map of (loanId, periodData). There may be multiple rows in the ResultSet and these must therefore be merged to be part of the same entry in the map.

The previous code does it by:

1. Read one LoanId from resultset and create a map entry
2. In a loop, keep fetching next row from result set. 
    - If the next row has the same loanId, then add it to the same map entry and fetch next
    - If the next row has a different loanId, go back one row in result set and start from step 1

The problem is the "go back one row" which is not supported by some JDBC drivers.

But an equivalent way of building the same map (at least in my view would be):

1. Iterate through the result set. For each row's loanId: 
- If we already have a map entry for that loanId, add the dates to the existing entry
- If we don't yet have a map entry for that loanId, create a new one and add the dates to that

No rs.previous() required, and either way you should end up with the same Map - as far as I can see..

But I don't have a way to test this code, so a bit concerned I may have missed something...

Reply:
Totally makes sense now. Thanks for helping me understand this better.
I actually read a comment that suggested doing this or using caching or hashmap or any other collection to hold all the resultset and instead iterate through it as you did but it just wasn't clear to me.

Unfortunately we don't have a way testing this but I agree with your approach. Since it passes all IT tests, we can just blindly merge this 2 or 3 days if there are no objections and hope for the best 😁.

Reply:
@ptuomola sorry for delay in reviewing this (I wanted to do it with a clear mind instead of clicking through it half asleep a night after day job work...) and after staring at this a little bit, I get it now - this makes good sense to me as well, and I have no objection to merging this even without having any way to functionally test this myself either (no idea if existing ITs, recently re-enable, would catch this, but doesn't matter, it "makes sense"). 

One very small feedback I would have is that, now that I understand this logic, it seems to me that moving that `scheduleDate.put(loanId, periodDatas);` now in line 449 (originally 434) up into the `if(periodDatas == null)` right after the `periodDatas = new ArrayList<>();` would make it a tiny bit more clear and explicit what we're doing here? It's two minor to hold back merging this important fix, so I'll just go ahead and merged, and let you raise a very small minor follow-up PR, if you like (or not, fine).